program/src/payload.rs

```diff
@@ -29,30 +26,90 @@ impl LeafInfo {
 
 #[repr(C)]
 #[cfg_attr(feature = "serde-feature", derive(Serialize, Deserialize))]
-#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Debug, Clone, Default)]
-pub struct Payload {
-    #[cfg_attr(
-        feature = "serde-feature",
-        serde(with = "As::<Option<DisplayFromStr>>")
-    )]
-    pub target: Option<Pubkey>,
-    pub derived_key_seeds: Option<SeedsVec>,
-    pub amount: Option<u64>,
-    pub tree_match_leaf: Option<LeafInfo>,
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Debug, Clone)]
+pub enum PayloadType {
+    Pubkey(Pubkey),
+    Seeds(SeedsVec),
+    MerkleProof(LeafInfo),
+    Number(u64),
 }
 
+#[repr(C)]
+#[cfg_attr(feature = "serde-feature", derive(Serialize, Deserialize))]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Debug, Clone, Default)]
+pub struct Payload(HashMap<PayloadKey, PayloadType>);
```

Comment:
Proposing that instead of using a Vec we use a HashMap that was defined as `HashMap<PayloadKey, PayloadType>`.  In order to make Auth Rules efficient, @blockiosaurus was already doing a very similar conversion inside of Auth Rules, since searching the Vec each time for a needed value was not an efficient approach.

This is the current Vec way we talked about with the team:
```
// Original from client:
let payload = vec![
    PayloadField::Authority(PayloadType::Pubkey(PUBKEY_A)),
    PayloadField::Amount(3)
];

// Asset manager instruction additions:
payload.push(PayloadField::Target(PayloadType::Pubkey(PUBKEY_B)));
payload.push(PayloadField::Amount(5));
In this case, Auth Rules gets two Amounts, and we don't know which one came from token-metadata.
```
Using a HashMap would instead be like this:
```
// Original from client:
let payload = Hashmap::from([(
    PayloadKey::Authority, PayloadType::Pubkey(PUBKEY_A)),
    (PayloadKey::Amount, PayloadType::Number(3)
]);

// Asset manager instruction additions:
payload.insert(PayloadKey::Target, PayloadType::Pubkey(PUBKEY_B));
payload.insert(PayloadKey::Amount, PayloadType::Number(5));
```
With the HashMap, the asset manager instruction overwrites `Amount` with the correct value and there's no ambiguity.  If you want to instead fail in this case, you would do `payload.try_insert()` which fails if key already present in HashMap.

The HashMap-based payload is still Vec-like in that the client can still pass in a payload and it can be modified or appended to by the asset manager.  There's some discussion as to whether there will be a case for the client to pass something in that the asset manager doesn't know about, but I don't think that changes the value proposition of this interface very much.

Overall a HashMap leads to more defined behavior by not allowing duplicates and requiring Auth Rules to know how to handle them. In my opinion it is better to leave that handling up to the Auth Rules caller.

Reply:
I like this.